crates/codegen/syntax_templates/src/shared/cursor.rs

```diff
@@ -0,0 +1,342 @@
+struct CursorPathElement {
+    rule_node: Rc<RuleNode>,
+    child_number: usize,
+    text_offset: TextIndex,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct CursorLeaf {
+    node: Node,
+    child_number: usize,
+    text_offset: TextIndex,
+}
+
+impl CursorLeaf {
+    pub fn text_range(&self) -> TextRange {
+        let start = self.text_offset;
+        let end = start + self.node.text_len();
+        start..end
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Cursor {
+    path: Vec<CursorPathElement>,
+    leaf: Option<CursorLeaf>,
+}
+
+impl Cursor {
+    pub(crate) fn new(node: Node) -> Self {
+        Self {
+            path: vec![],
+            leaf: Some(CursorLeaf {
+                node,
+                child_number: 0,
+                text_offset: Default::default(),
+            }),
+        }
+    }
+
+    // Unlike clone, this re-roots at the current node.
+    // It does preserve the correct text offset however,
+    // even though the path is reset.
+    pub fn spawn(&self) -> Self {
+        Self {
+            path: vec![],
+            leaf: self.leaf.clone(),
+        }
+    }
+
+    pub fn is_finished(&self) -> bool {
+        self.leaf.is_none()
+    }
+
+    pub fn node_or_none(&self) -> Option<Node> {
+        self.leaf.as_ref().map(|leaf| leaf.node.clone())
+    }
+
+    pub fn node(&self) -> Node {
+        self.leaf
+            .as_ref()
+            .expect("This cursor is finished, so it has no node.")
+            .node
+            .clone()
+    }
+
+    pub fn text_offset(&self) -> TextIndex {
+        self.leaf
+            .as_ref()
+            .expect("This cursor is finished, so it has no text_offset`")
+            .text_offset
+    }
+
+    pub fn text_range(&self) -> TextRange {
+        self.leaf
+            .as_ref()
+            .expect("This cursor is finished, so it hasn no text_range")
+            .text_range()
+    }
+
+    pub fn path_rule_nodes(&self) -> Vec<Rc<RuleNode>> {
+        self.path
+            .iter()
+            .map(|path_element| path_element.rule_node.clone())
+            .collect()
+    }
+
+    pub fn go_to_next(&mut self) -> bool {
+        if !self.go_to_first_child() {
+            return self.go_to_next_non_descendent();
+        }
+        return true;
+    }
+
+    pub fn go_to_next_non_descendent(&mut self) -> bool {
+        while !self.go_to_next_sibling() {
+            if !self.go_to_parent() {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    pub fn go_to_previous(&mut self) -> bool {
+        while !self.go_to_previous_sibling() {
+            if !self.go_to_parent() {
+                return false;
+            }
+        }
+        while self.go_to_last_child() {}
+        return true;
+    }
+
+    pub fn go_to_parent(&mut self) -> bool {
+        self.leaf = self.path.pop().map(|path_element| CursorLeaf {
+            node: Node::Rule(path_element.rule_node),
```

Comment:
IIUC, that means that when there is an empty path with an existing leaf, this method will delete the leaf and return false. If there is no parent (path is empty), shouldn't we return false without modifying the current cursor?

Reply:
I went back and forth about this. It is assumed in methods that call `go_to_parent()` that failure of that method also completes the cursor. I didn't like that one navigator having different behaviour. Since you raised it, I am swayed and I will add an explicit `complete()` method in the callers.

Reply:
Actually, now that I think about it, it needs to be like this. The `go_to_*` methods fall into two categories:
1. Those that fail without moving, which are only the `go_to_*_sibling` methods; and
2. Those that fail and complete the cursor in the process. Which is all others - and those all fail via `go_to_parent`, which is the only place that completes the cursor.

We have to complete the cursor so that the start and end states are distinguishable, otherwise the cursor becomes cyclic. I've built cursor APIs like that before, and always regretted it. The DX is awful.

Reply:
now more obvious, with explicit completion